### PR TITLE
feat(qkernel): auto-promote scalar literals at sub-qkernel call sites

### PR DESCRIPTION
**Pull request by ksk-jij**
## Summary

- Sub-`@qkernel` calls now accept raw Python literals (`int`/`float`/`bool`) for `UInt`/`Float`/`Bit` parameters — `ry_layer(q, thetas, 0)` works the same as `ry_layer(q, thetas, qmc.uint(0))`.
- Mirrors the literal-acceptance pattern of built-in gate primitives like `qmc.rx`. As a beneficial side effect, `def f(n: UInt = 4)` defaults now work without explicit wrapping.
- Already-Handle args and array-typed params pass through unchanged — existing symbolic-execution paths are unaffected (verified by 6118 passing tests).

## Motivation

A user reported that the natural HEA pattern below raises `TypeError: Argument 'offset' must be a Handle instance, got <class 'int'>`:

\`\`\`python
@qmc.qkernel
def hea(n: qmc.UInt, thetas: qmc.Vector[qmc.Float]) -> qmc.Vector[qmc.Bit]:
    q = qmc.qubit_array(n, name=\"q\")
    q = ry_layer(q, thetas, 0)        # ← had to wrap as qmc.uint(0)
    q = cz_entangling_layer(q)
    q = ry_layer(q, thetas, n)
    return qmc.measure(q)
\`\`\`

The strict isinstance check at \`QKernel.__call__\` rejected raw literals even though the callee's signature unambiguously declared the expected scalar type.

## Implementation

- Added \`_promote_literal_to_handle(value, expected_type)\` in \`qamomile/circuit/frontend/qkernel.py\` that wraps raw \`int\`/\`float\`/\`bool\` literals via the existing \`qmc.uint\`/\`qmc.float_\`/\`qmc.bit\` constructors when the callee's declared annotation is exactly \`UInt\`/\`Float\`/\`Bit\`.
- Hooked into \`QKernel.__call__\` immediately after \`apply_defaults()\`, so both call-site literals and scalar default values are wrapped on the same path.
- \`bool\` is explicitly excluded from \`int → UInt\` / \`int → Float\` to avoid silent \`True → UInt(1)\` coercion. \`bool\` is only ever promoted to \`Bit\`.

## Tests

New \`tests/circuit/test_qkernel_literal_promotion.py\` covers 40 cases:

- **UInt/Float/Bit literal promotion** (Qiskit + QuriParts + CUDA-Q × sample + expval): the user's HEA case, analytic-baseline checks (\`<H> = 2 cos θ\`), and cross-checks against explicitly-wrapped variants.
- **Negative cases**: \`str → Float\`, \`list → Vector[Float]\`, \`bool → UInt\`, \`str → Bit\` all keep raising the original \`TypeError\`. The \`bool → UInt\` test specifically pins the design choice that \`bool\` is excluded from numeric promotion.
- **Default-value edge case**: \`def f(n: UInt = 4)\` is now callable from a kernel body without wrapping the default — verified across all 3 backends via H-broadcast sampling.

Existing suites: \`tests/circuit/\` + \`tests/transpiler/\` all-markers run **6118 passed, 0 failed**.

## Documentation

Added a \"Passing scalar literals to helpers / ヘルパーへのスカラーリテラルの受け渡し\" subsection to \`docs/{en,ja}/tutorial/06_reuse_patterns.py\` (and regenerated \`.ipynb\`) demonstrating the new behavior next to the helper-qkernel example.

## Test plan

- [x] New tests pass on Qiskit + QuriParts + CUDA-Q (sample + expval)
- [x] Full \`tests/circuit/\` + \`tests/transpiler/\` suite passes (6118 / 0)
- [x] Tutorial 06 docs tests pass (en + ja)
- [x] ruff format / check clean

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tests/circuit/test_qkernel_literal_promotion.py [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+        assert np.isclose(out_lit, out_wrap, atol=1e-6)
+
+
+# ---------------------------------------------------------------------------
+# Bit literal promotion
+# ---------------------------------------------------------------------------
+#
+# Less common, but the symmetry argument applies: a callee declaring
+# ``flag: qmc.Bit`` should accept ``True``/``False``/``0``/``1`` at the
+# call site. Compile-time folding then erases the dead branch.
```
[the remark this answers is not in the document]
Comment (ksk-jij) [reply tests/circuit/test_qkernel_literal_promotion.py]: Fixed in b3394f28. The section comment overstated the supported set — the implementation deliberately accepts only `bool` (`True`/`False`) so the rule is symmetric to `bool` being excluded from the `int → UInt` and `int → Float` paths: each scalar Handle accepts only its semantic Python primitive. I updated the comment to match, and added `test_int_for_bit_param_raises` to lock the `int → Bit` rejection alongside the existing `bool → UInt` rejection test. Users who want a Bit-typed integer literal can wrap it explicitly via `qmc.bit(0)`.

File: tests/circuit/test_qkernel_literal_promotion.py [hunk 2]
```diff
@@ -0,0 +1,494 @@
...
+except ImportError:  # pragma: no cover - covered when cudaq is absent
+    _HAS_CUDAQ = False
+    CudaqTranspiler = None  # type: ignore[assignment]
+
+_HAS_OBSERVABLE = True
+try:  # pragma: no cover - presence check, not behaviour
+    import qamomile.observable as qm_o
+except ImportError:  # pragma: no cover - never expected, defensive
+    _HAS_OBSERVABLE = False
+    qm_o = None  # type: ignore[assignment]
```
[the remark this answers is not in the document]
Comment (ksk-jij) [reply tests/circuit/test_qkernel_literal_promotion.py]: Fixed in b3394f28. You're right — `qamomile.observable` is part of the project itself, not an optional SDK, so the try/except was dead defensive code that would have masked a broken install rather than skipping cleanly: `qm_o = None` would have crashed in `_sum_z_hamiltonian` exactly as you noted. Removed the guard and made the import unconditional alongside the other in-project imports.

File: qamomile/circuit/frontend/qkernel.py [hunk 3]
```diff
@@ -54,6 +54,54 @@ def _get_array_element_type(pt: Any) -> type | None:
...
+    if isinstance(value, Handle):
+        return value
+    is_bool = isinstance(value, bool)
+    if expected_type is UInt:
+        if isinstance(value, int) and not is_bool:
+            return uint(value)
+    elif expected_type is Float:
+        if isinstance(value, float) or (isinstance(value, int) and not is_bool):
+            return float_(float(value))
+    elif expected_type is Bit:
```
[the remark this answers is not in the document]
Comment (ksk-jij) [reply qamomile/circuit/frontend/qkernel.py:107]: Fixed in 249c8caf. The `int → Float` promotion is intentional — it mirrors Python's natural `int → float` numeric coercion (e.g., `math.cos(0)` returns `1.0`), so it would be surprising for `Float` to be the only place that rejects it. To prevent regressions I added `test_int_literal_promotes_to_float_param` (4 thetas × 3 backends) that locks the rule via the closed-form expval ⟨Z₀+Z₁⟩ = 2 cos(θ) on `int` inputs. I also tightened the `_promote_literal_to_handle` docstring to enumerate the per-type rules explicitly (`UInt` ← int; `Float` ← int|float; `Bit` ← bool only) so the contract is unambiguous.

File: tests/circuit/test_qkernel_literal_promotion.py [hunk 4]
```diff
@@ -0,0 +1,512 @@
...
+        out_uint = dict(
+            exe_uint.sample(t.executor(), shots=shots, bindings=bindings_runtime)
+            .result()
+            .results
+        )
+        # Same set of measured bitstrings (counts may differ by shot noise).
+        assert set(out_int.keys()) == set(out_uint.keys()), (
+            f"[{transpiler_factory.__name__}] outcome support differs: "
+            f"int-literal={set(out_int.keys())}, qmc.uint(0)={set(out_uint.keys())}"
+        )
```
[the remark this answers is not in the document]
Comment (ksk-jij) [reply tests/circuit/test_qkernel_literal_promotion.py]: Fixed in 249c8caf. You're right — the set-equality assertion would miss low-probability bitstrings that one run happens to skip on a non-uniform HEA distribution. I replaced it with an empirical total-variation-distance bound (TVD < 0.1). For 4096 shots over 8 outcomes, per-outcome shot-noise std is ≤ 0.008, so a 0.1 threshold is comfortably above sampling noise across independent RNG seeds while still catching real divergence. The test was renamed accordingly (`test_int_literal_sampling_matches_qmc_uint_distribution`). Strict numerical equivalence between the two forms remains tested separately via expval in `test_int_literal_offset_expval_matches_analytic`.
